src/components/Swap/SwapButton/index.tsx

```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
-      addTransaction({
-        response: transaction,
-        type: wrapType,
-        amount: CurrencyAmount.fromRawAmount(native, transaction.value?.toString() ?? '0'),
-      })
-      setDisplayTxHash(transaction.hash)
-      // Only reset pending after any queued animations to avoid layout thrashing, because a
-      // successful wrap will open the status dialog and immediately cover the button.
-      const postWrap = () => {
-        setIsPending(false)
-        document.removeEventListener('animationend', postWrap)
-      }
-      if (isAnimating(document)) {
-        document.addEventListener('animationend', postWrap)
-      } else {
-        setIsPending(false)
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from wrap.
-      console.log(e)
-    } finally {
-      setIsPending(false)
-    }
-  }, [addTransaction, native, setDisplayTxHash, wrapCallback, wrapType])
+    await onSubmit(async () => {
+      const response = await wrapCallback()
+      if (!response) return
+
+      invariant(wrapType)
+      const amount = CurrencyAmount.fromRawAmount(native, response.value?.toString() ?? '0')
+      return { response, type: wrapType, amount }
+    })
+
+    // Whether or not the transaction submits, we should still reset the pending state.
+    setIsPending(false)
+  }, [native, onSubmit, wrapCallback, wrapType])
   // Reset the pending state if user updates the swap.
   useEffect(() => setIsPending(false), [inputCurrencyAmount, trade])
 
+  const setOldestValidBlock = useSetOldestValidBlock()
   const onSwap = useCallback(async () => {
-    try {
-      const transaction = await swapCallback?.()
-      if (!transaction) return
-      invariant(trade.trade)
-      addTransaction({
-        type: TransactionType.SWAP,
-        response: transaction,
-        tradeType: trade.trade.tradeType,
-        trade: trade.trade,
-      })
-      setDisplayTxHash(transaction.hash)
+    const submitted = await onSubmit(async () => {
+      const response = await swapCallback?.()
+      if (!response) return
 
       // Set the block containing the response to the oldest valid block to ensure that the
       // completed trade's impact is reflected in future fetched trades.
-      transaction.wait(1).then((receipt) => {
+      response.wait(1).then((receipt) => {
         setOldestValidBlock(receipt.blockNumber)
       })
 
-      // Only reset open after any queued animations to avoid layout thrashing, because a
-      // successful swap will open the status dialog and immediately cover the summary dialog.
-      const postSwap = () => {
-        setOpen(false)
-        document.removeEventListener('animationend', postSwap)
-      }
-      if (isAnimating(document)) {
-        document.addEventListener('animationend', postSwap)
-      } else {
-        setOpen(false)
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from swap.
-      console.log(e)
+      invariant(trade.trade)
```

Comment:
is adding an `invariant` call here the correct move? iirc it throws if the argument is falsy, so this would be an uncaught error and trigger the error boundary.

Reply:
`onWrap` will not be called if `wrapType === undefined`, but due to the complex type that's not evident to the type checker. We _could_ replace this with a check, but if `wrapCallback()` returns a valid response, then `wrapType` should also be valid. I can annotate this with a comment.